public/lesswrong-editor/src/ckeditor5-footnote/src/footnoteEditing/googleDocsFootnotesNormalizer.js

```diff
@@ -0,0 +1,261 @@
+		});
+		return [...documentRange]
+		.reduce((acc, { item, type }) => {
+			if (
+				type === 'elementStart' &&
+				footnoteReferenceMatcher.match(item)
+			) {
+				// ensure that a matching footnote exists for this reference
+				// @ts-ignore: matcher above eliminates the null / undefined cases here.
+				const index = item.getAttribute('id').match(idPattern)[1]
+				if(backLinks.hasOwnProperty(index)) {
+					return {...acc, [index]: { item, id: backLinks[index] }};
+				}
+			}
+			return acc;
+		}, {});
+	}
+
+	/**
+	 * For each unprocessed backLink, creates a new element composed of a
+	 * footnoteBackLink element and a footnoteContent element, all wrapped in a footnoteItem.
+	 *
+	 * Presumes a structure s.t. each backlink is wrapped in two elements, the outer of which
+	 * contains the entire contents of the footnote.
+	 * @param {UpcastWriter} writer
+	 * @param {Object.<string, {item: Element, id: string}>} backLinks 
+	 * @returns {Element[]}
+	 */
+	_createFootnoteItems(writer, backLinks) {
+		return Object.entries(backLinks).map(([ index, {item, id}]) => {
+			if(!item.parent || !item.parent.parent || !item.parent.parent.is('element')) {
+				throw new Error("Unexpected Dom structure; expected footnote backLink to be nested within two parent tags.")
+			}
+			const footnoteContent = item.parent.parent;
+			/**
+			 * These range bounds (start after the backLink, end
+			 * after the footnote content) are used to capture
+			 * all of the footnote except the backLink, which we
+			 * generate ourselves.
+			 * @type {Range}
+			*/
+			// @ts-ignore DefinitelyTyped mistyping of Range
+			const footnoteContentRange = writer.createRange(
+				writer.createPositionAfter(item),
+				writer.createPositionAfter(footnoteContent),
+			)
+			const newFootnoteBackLink = writer.createElement('span', {
+				[ATTRIBUTES.footnoteBackLink]: '',
+				[ATTRIBUTES.footnoteId]: id,
+			});
+			const newFootnoteContent = writer.createElement('div', 
+				{
+					[ATTRIBUTES.footnoteContent]: '',
+					[ATTRIBUTES.footnoteId]: id,
+					[ATTRIBUTES.footnoteIndex]: index,
+				},
+
+				// @ts-ignore: startPosition isn't required here.
+				[...footnoteContentRange.getItems({ shallow: true})]
+					.filter(item => item.is('element'))
+					// @ts-ignore: previous line is a type guard.
+					.map(element => writer.clone(element, true)),
+			);
+			return writer.createElement('div', { 
+				[ATTRIBUTES.footnoteItem]: '',
+				[ATTRIBUTES.footnoteId]: id,
+				[ATTRIBUTES.footnoteIndex]: index,
+			}, [newFootnoteBackLink, newFootnoteContent]);
+		});
+	}
+
+	/**
+	 * Adds `footnoteReference` `footnoteIndex`, and `footnoteId`
+	 * attributes to each footnote reference.
+	 * @param {UpcastWriter} writer
+	 * @param {Range} documentRange
+	 * @param {Object.<string, {item: Element, id: string}>} backLinks
+	 */
+	_preprocessFootnoteReferences(writer, documentRange, backLinks) {
+		const references = this._getFootnoteReferences(documentRange, backLinks);
```

Comment:
Actually, I honestly think this would be clearer to the reader of `execute` if this line were in `execute` and `references` was passed in as an argument.

🌌 🧠

Reply:
Thanks, agreed!